### ISSUE TITLE
[Feature Request] Example or Record class support for Dataset

### ISSUE THREAD

**Issue by DNGros** [pytorch/pytorch#16951]
**(Accidentally hit enter while writing and can't delete. Working on this)**

**Comment by DNGros** [comment]
Sorry, accidentally hit enter or clicked or something and published this issue at very beginning of writing this up. As I dug into the internals of dataloader and dataset some more, I realized what I originally had in mind isn't the best idea. 

This was initially spawned after frustration with not being able to return attr.s/Python-3.7-Dataclass-like-objects in a Dataset.__getitem__ without Dataloader being unhappy. I was also envisioning something similar to tf.train.Example, as there does not seem to be a clear analog in pytorch.

After working with this more I might open a new more coherent issue. Sorry for this spam issue!

**Comment by fmassa** [comment]
@DNGros note that you can have a `Dataclass` without problems, but you might need to pass a custom `collate_fn` to the dataloader

**Comment by DNGros** [comment]
@fmassa , thanks for pointing that out. I had noticed the `collate_fn` parameter after going through the code a bit more (which was one reason for abandoning the original issue).

A possible improvement might be that for anything that is an instance of `TorchRecord` (or some name) the `default_collate` would recursively call itself all the `TorchRecord`'s attributes, similar to the recursive behavior for each items in dicts.

As a temporary solution right now though I'm not using `Dataloader` at all and I'm just manually iterating over my data. This is because it was quicker to do before I realized about the custom collate, and because of a separate issue where even with a custom `collate_fn` there doesn't seem to be a nice way of yielding batches using some custom selection function, like for example to group together similar sequence lengths to minimize padding. torchtext's BucketIterator provides a single-process example of this. However, the multiprocessing functionality of `Dataloader` is really cool, so hopefully it'll eventually be possible to get it all working together.
